src/expr/node_manager_template.cpp

```diff
@@ -615,6 +615,10 @@ std::vector<TypeNode> NodeManager::mkMutualDatatypeTypesInternal(
   DatatypeIndexAttr dia;
   for (const DType& dt : datatypes)
   {
+    if (dt.isResolved())
+    {
+      throw Exception("datatype is already resolved");
```

Comment:
Why throw an `Exception` instead of making this an assertion? If we throw an exception, it should maybe be an `IllegalArgumentException`?

Reply:
I was following the other exception thrown in this method: https://github.com/cvc5/cvc5/blob/main/src/expr/node_manager_template.cpp#L657

Should this change? What is the difference between these exceptions?

Reply:
I was initially planning to comment about this too, but then I looked more into this. I think we don't want to introduce any new IllegalArgumentExceptions (and no PrettyCheckArgument and friends). Those were originally used for the guards in the old API, and none of these should be exceptions but assertions (there might be some we have to convert). I guess we should only have ModalException and Exception (and the recoverable versions). Now, the question if we want to throw exceptions in the node manager at all is a question we should discuss in more detail, and are probably out of scope for this PR.

Reply:
After an offline discussion, we agreed to keep the exception here for now, and will make an effort to convert all exceptions in node_manager_template to assertions where possible.